Allow users to differentiate task pods from different namespaces in multi-namespace settings

GWphua opened the pull request:
### Description

Related to #17738. 

When we have task pods running in multiple namespaces, it will be difficult to tell which task pods are being started by which cluster, especially when we have very similar job names. 
 
#### Added an alias prefix tag to allow Kubernetes operators to identify which cluster the task pods are created from.
Here are some of the things I have considered when implementing the change:
1. The job name must respect the 63-character limit.
2. We can easily differentiate between tasks and namespaces.
3. The job name must be unique.
4. What if users are running everything within one namespace? Maybe users will still prefer the old way of displaying task ID.

Intuition:
`k8sJobName` is currently broken down into 2 segments: {taskId}-{hash(taskId)}. The hashing will take up 32 characters, and is needed to maintain uniqueness. Hence, we can only work with making changes to the `taskId` part. The first thought is to change `taskId` to namespace, so that we at least know which namespace a task pod belongs in. However, I encountered cases where namespace have long names, and similar prefix (Think `data-infra-druid-cluster-ns1`, `data-infra-druid-cluster-ns2`).  This is not a good solution, given that there's a character limit to the job name and there is quite a lot of redundant information.

So, I introduced a new `druid.indexer.runner. k8sTaskPodNamePrefix ` config to allow users to provide an alias prefix instead. Now, `k8sJobName` will be of the form {alias}-{hash(taskId)}. If `k8sTaskPodNamePrefix` is not provided, we will default to using the current {taskId}-{hash(taskId)} naming convention.

Note that this solution will not allow us to differentiate between tasks in Kubernetes. However, we can still see our tasks and their corresponding taskID via the Web Console.

Note: My approach involves changing the method of K8sTaskId, prompting change in a lot other classes. This may not be the best way, please advise if there's a more elegant solution.

#### Release note
You can now name your K8s job names using `druid.indexer.runner. k8sTaskPodNamePrefix`

<hr>

##### Key changed/added classes in this PR
 * KubernetesPeonLifecycle.java
 * KubernetesPeonLifecycleFactory.java
 * KubernetesTaskRunner.java
 * KubernetesTaskRunnerConfig.java
 * PeonLifecycleFactory.java
 * K8sTaskId.java
 * KubernetesOverlordUtils.java
 * K8sTaskAdapter.java
 * MultiContainerTaskAdapter.java
 * PodTemplateTaskAdapter.java
 * SingleContainerTaskAdapter.java

##### Test Files
 * KubernetesPeonLifecycleTest.java
 * KubernetesWorkItemTest.java
 * TestPeonLifecycleFactory.java
 * K8sTaskIdTest.java
 * KubernetesPeonClientTest.java
 * DruidPeonClientIntegrationTest.java
 * K8sTaskAdapterTest.java
 * PodTemplateTaskAdapterTest.java


<hr>

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

FrankChen021 commented:
the name `druid.indexer.runner.alias` is not intuitive. Maybe `druid.indexer.runner.k8sTaskNamePrefix` ?
and please add document for this new configuration.

GWphua commented:
hi @FrankChen021, I will work on documentation and changes after #17738 is merged. Since this PR will not make sense without it.

GWphua commented:
> the name `druid.indexer.runner.alias` is not intuitive. Maybe `druid.indexer.runner.k8sTaskNamePrefix` ? and please add document for this new configuration.

Hi @FrankChen021, I have changed the name to `druid.indexer.runner.k8sTaskPodNamePrefix`. Also, added the documentation for this new configuration. Feel free to take a look.

FrankChen021 reviewed (approved):
LGTM with a minor document description suggestion

FrankChen021 commented on extensions-core/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerConfig.java:
```diff
@@ -371,6 +382,12 @@ public Builder withOverlordNamespace(String overlordNamespace)
       return this;
     }
 
+    public Builder withk8sTaskPodNamePrefix(String k8sTaskPodNamePrefix)
```
```suggestion
    public Builder withK8sTaskPodNamePrefix(String k8sTaskPodNamePrefix)
```

suggested by Copilot

kfaraz commented:
> Note that this solution will not allow us to differentiate between tasks in Kubernetes. However, we can still see our tasks and their corresponding taskID via the Web Console.
Note: My approach involves changing the method of K8sTaskId, prompting change in a lot other classes. This may not be the best way, please advise if there's a more elegant solution.

Yeah, @GWphua , I am not sure if this is a real improvement over what we had earlier.
With a prefix specified, all task pods launched by a cluster will end up looking the same (except for the hash part, which is not exactly human readable). It would be difficult to identify the pod for any given Druid task ID.
Plus, requiring users to specify a prefix just to be able to tell pods apart seems undesirable to me.

That said, using the substring of the task name was not all that great either.
I am not sure what a better alternative would be, but will try to give it some more thought.

Side note:
IIUC from the changes, this shouldn't have any rolling upgrade implications, i.e. task pods launched by an Overlord where the `k8sTaskPodNamePrefix` was not specified can still be detected by an Overlord where this config has been specified (or an Overlord where the value of the config has been updated).
@GWphua , could you please confirm if this is actually the case.

GWphua commented:
> IIUC from the changes, this shouldn't have any rolling upgrade implications, i.e. task pods launched by an Overlord where the k8sTaskPodNamePrefix was not specified can still be detected by an Overlord where this config has been specified (or an Overlord where the value of the config has been updated).
@GWphua , could you please confirm if this is actually the case.

Hi @kfaraz, unfortunately there will be disruptions to rolling upgrades. 

The old tasks (without configured `k8sTaskPodNamePrefix`) will not be detected, and Overlord will restart the on-going tasks with the configured `k8sTaskPodNamePrefix`. From the user's perspective, the web console will show the old tasks as FAILED.

On our side, we need to notify our users to expect this to happen.

kfaraz commented:
> The old tasks (without configured k8sTaskPodNamePrefix) will not be detected, and Overlord will restart the on-going tasks with the configured k8sTaskPodNamePrefix. From the user's perspective, the web console will show the old tasks as FAILED.

Oh, this is problematic, @GWphua , as the Overlord is losing track of those tasks and re-launching them, even though the original tasks are still running and will finish normally. This will lead to duplicate data and possibly metadata inconsistency errors in case of streaming ingestion.

Is there any way to mitigate this?

kfaraz commented:
> Currently I think we can add some text to the doc to give a guidance on changing of this property which requires users to stop all running tasks first.

Yes, let's update the docs and we should also clearly call this out in the release notes and upgrade notes for this feature (just adding it in the PR description should suffice).

cc: @GWphua
